yoti-sdk-impl/src/test/java/com/yoti/api/client/spi/remote/DataEntryConverterTest.java

```diff
@@ -0,0 +1,82 @@
+
+import java.text.ParseException;
+
+import static org.hamcrest.Matchers.containsString;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class DataEntryConverterTest {
+
+    private static final String SOME_DATA_ENTRY = "someBytes";
+    private static final ByteString SOME_DATA_ENTRY_BYTE_STRING = ByteString.copyFromUtf8(SOME_DATA_ENTRY);
+
+    @InjectMocks
+    DataEntryConverter dataEntryConverter;
+
+    @Mock
+    ThirdPartyAttributeConverter thirdPartyAttributeConverterMock;
+    @Mock
+    DataEntryProto.DataEntry dataEntryMock;
+    @Mock
+    CredentialIssuanceDetails credentialIssuanceDetailsMock;
+
+    @Test
+    public void shouldParseDataEntryCorrectly() throws Exception {
+        when(dataEntryMock.getType()).thenReturn(DataEntryProto.DataEntry.Type.THIRD_PARTY_ATTRIBUTE);
+        when(dataEntryMock.getValue()).thenReturn(SOME_DATA_ENTRY_BYTE_STRING);
+        when(thirdPartyAttributeConverterMock.parseThirdPartyAttribute(SOME_DATA_ENTRY_BYTE_STRING.toByteArray())).thenReturn(credentialIssuanceDetailsMock);
+
+        dataEntryConverter.convertDataEntry(dataEntryMock);
+    }
+
+    @Test
+    public void shouldThrowExceptionForUnsupportedProtobufType() {
+        when(dataEntryMock.getType()).thenReturn(DataEntryProto.DataEntry.Type.UNDEFINED);
+
+        try {
+            dataEntryConverter.convertDataEntry(dataEntryMock);
+        } catch (ExtraDataException e) {
+            assertThat(e.getMessage(), containsString("Unsupported data entry"));
+            return;
+        }
+        fail("Expected ExtraDataException");
+    }
+
+    @Test(expected = ExtraDataException.class)
+    public void shouldWrapInvalidProtocolBufferException() throws Exception {
+        InvalidProtocolBufferException invalidProtocolBufferException = new InvalidProtocolBufferException("Some Exception");
+
+        when(dataEntryMock.getType()).thenReturn(DataEntryProto.DataEntry.Type.THIRD_PARTY_ATTRIBUTE);
+        when(dataEntryMock.getValue()).thenReturn(SOME_DATA_ENTRY_BYTE_STRING);
+        when(thirdPartyAttributeConverterMock.parseThirdPartyAttribute(SOME_DATA_ENTRY_BYTE_STRING.toByteArray())).thenThrow(invalidProtocolBufferException);
+
+        dataEntryConverter.convertDataEntry(dataEntryMock);
+    }
+
+    @Test(expected = ExtraDataException.class)
+    public void shouldWrapParseException() throws Exception {
+        ParseException parseException = new ParseException("Some exception", 0);
+
+        when(dataEntryMock.getType()).thenReturn(DataEntryProto.DataEntry.Type.THIRD_PARTY_ATTRIBUTE);
+        when(dataEntryMock.getValue()).thenReturn(SOME_DATA_ENTRY_BYTE_STRING);
+        when(thirdPartyAttributeConverterMock.parseThirdPartyAttribute(SOME_DATA_ENTRY_BYTE_STRING.toByteArray())).thenThrow(parseException);
+
+        dataEntryConverter.convertDataEntry(dataEntryMock);
+    }
```

Comment:
Same here - checking that this _isn't_ a ParseException?

Reply:
Same again, in the `@Test` annotation we're expecting an `ExtraDataException` to be thrown

Reply:
Same for this one, all good then